Sources/ClientRuntime/Networking/Streaming/BufferedStream.swift

```diff
@@ -0,0 +1,124 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// A `Stream` implementation that buffers data in memory.
+/// The buffer size depends on the amount of data written and read.
+/// Note: This class is thread-safe.
+public class BufferedStream: Stream {
+
+    /// The length of the stream, if known.
+    public var length: Int?
+
+    /// The current position in the stream.
+    public var position: Data.Index
+
+    /// Whether the stream is empty.
+    public var isEmpty: Bool {
+        return buffer.isEmpty
+    }
+
+    /// Whether the stream is seekable.
+    /// Although this stream is seekable, seeking to a position that is not in the buffer will cause the stream to throw an error.
+    public var isSeekable: Bool = false
+
+    var isClosed: Bool
+
+    var buffer = Data()
+    let lock = NSLock()
```

Comment:
Can we make this private?

Reply:
done

Reply:
perhaps not buffer (tests)  but made if private(set)